packages/node/src/lib/token-manager.ts

```diff
@@ -0,0 +1,315 @@
+    this.tokenEmitter.on('access_token', (event) => {
+      if ('token' in event) {
+        this.accessToken = event.token
+      }
+    })
+    this.retryCount = 0
+  }
+
+  async pollerLoop() {
+    let timeUntilRefreshInMs = 25
+    let response: HTTPResponse
+    let headers: Record<string, string>
+
+    try {
+      response = await this.requestAccessToken()
+      headers = convertHeaders(response.headers)
+    } catch (err) {
+      // Error without a status code - likely networking, retry
+      this.retryCount++
+      this.lastError = err
+
+      if (this.retryCount % this.maxRetries === 0) {
+        this.tokenEmitter.emit('access_token', { error: this.lastError })
+      }
+
+      this.pollerTimer = setTimeout(
+        this.pollerLoop.bind(this),
+        backoff({
+          attempt: this.retryCount,
+          minTimeout: 25,
+          maxTimeout: 1000,
+        })
+      )?.unref()
+      return
+    }
+    if (headers && headers['date']) {
+      try {
+        this.clockSkewInSeconds =
+          (Date.now() - Date.parse(headers['date'])) / 1000
+      } catch (err) {
+        // Unable to parse, move on with last or 0 skew
+      }
+    }
+
+    // Handle status codes!
+    if (response.status === 200) {
+      let body: any
+      if (typeof response.text != 'function') {
+        this.tokenEmitter.emit('access_token', {
+          error: new Error(
+            'HTTPClient does not implement response.text method'
+          ),
+        })
+        return
+      }
+      try {
+        body = await response.text()
+      } catch (err) {
+        // Errors reading the body (not parsing) are likely networking issues, we can retry
+        this.retryCount++
+        this.lastError = err
+        timeUntilRefreshInMs = backoff({
+          attempt: this.retryCount,
+          minTimeout: 25,
+          maxTimeout: 1000,
+        })
+        this.pollerTimer = setTimeout(
+          this.pollerLoop.bind(this),
+          timeUntilRefreshInMs
+        )?.unref()
+        return
+      }
+      try {
+        const token = JSON.parse(body)
+
+        if (!isAccessToken(token)) {
+          throw new Error(
+            'Response did not contain a valid access_token and expires_in'
+          )
+        }
+
+        token.expires_at = Math.round(Date.now() / 1000) + token.expires_in
+
+        this.tokenEmitter.emit('access_token', { token })
+
+        // Reset our failure count
+        this.retryCount = 0
+
+        // Refresh the token after half the expiry time passes
+        timeUntilRefreshInMs = (token.expires_in / 2) * 1000
+      } catch (err) {
+        // Something went really wrong with the body, lets surface an error and try again?
+        this.tokenEmitter.emit('access_token', { error: err })
+        this.retryCount = 0 // Reset because we've emitted an error
+
+        timeUntilRefreshInMs = backoff({
+          attempt: this.retryCount,
+          minTimeout: 25,
+          maxTimeout: 1000,
+        })
```

Comment:
I think it's a little confusing to follow given how long is this function to recognize where this is being used (down below).

Maybe it would be worth splitting the paths of each error code into different functions to make it easier to follow:
- handleOkResponse()
- handleRateLimited()
- handleError()
- handleUnrecoverableError

Reply:
Did this for all of them except the OK response since that one seemed a little trickier, let me know if the changes helped or made things worse!